blockchain/vm/internaltx_tracer.go

```diff
@@ -0,0 +1,495 @@
+		this.callStack = []*InternalCall{{}}
+	}
+	result := &InternalTxTraceResult{
+		callType: this.ctx["type"].(OpCode),
+		from:     this.ctx["from"].(common.Address),
+		to:       this.ctx["to"].(common.Address),
+		value:    "0x" + this.ctx["value"].(*big.Int).Text(16),
+		gas:      this.ctx["gas"].(uint64),
+		gasUsed:  this.ctx["gasUsed"].(uint64),
+		input:    this.ctx["input"].(string),
+		output:   this.ctx["output"].(string),
+		time:     this.ctx["time"].(time.Duration),
+	}
+	if this.callStack[0].calls != nil {
+		result.calls = this.callStack[0].calls
+	}
+	if this.callStack[0].Err != nil {
+		result.error = this.callStack[0].Err
+	} else if ctxErr, _ := this.ctx["error"]; ctxErr != nil {
+		result.error = ctxErr.(error)
+	}
+	if result.error != nil {
+		result.output = "" // delete result.output;
+	}
+	if this.ctx["error"] == errEvmExecutionReverted {
+		outputHex := this.ctx["output"].(string) // it is already a hex string
+		if outputHex[2:10] == "08c379a0" {
+			defaultOffset := 10
+
+			stringOffset, err := strconv.ParseInt(outputHex[defaultOffset:defaultOffset+32*2], 16, 64)
+			if err != nil {
+				logger.Error("failed to parse hex string to get stringOffset",
+					"err", err, "outputHex", outputHex)
+				return nil, err
+			}
+			stringLength, err := strconv.ParseInt(outputHex[defaultOffset+32*2:defaultOffset+32*2+32*2], 16, 64)
+			if err != nil {
+				logger.Error("failed to parse hex string to get stringLength",
+					"err", err, "outputHex", outputHex)
+				return nil, err
+			}
+			start := defaultOffset + 32*2 + int(stringOffset*2)
+			end := start + int(stringLength*2)
+			asciiInBytes, err := hex.DecodeString(outputHex[start:end])
+			if err != nil {
+				logger.Error("failed to parse hex string to get ASCII representation",
+					"err", err, "outputHex", outputHex)
+				return nil, err
+			}
+			this.revertString = string(asciiInBytes)
+		}
+		result.reverted = revertedInfo{contract: this.revertedContract, message: this.revertString}
+	}
+	return result, nil
+}
+
+// InternalTxLogs returns the captured tracerLog entries.
+func (this *InternalTxTracer) InternalTxLogs() []*InternalCall { return this.callStack }
+
+// fault is invoked when the actual execution of an opcode fails.
+func (this *InternalTxTracer) fault(log *tracerLog) {
+	// If the topmost call already reverted, don't handle the additional fault again
+	if this.callStack[this.callStackLength()-1].Err != nil {
+		return
+	}
+	// Pop off the just failed call
+	call := this.callStackPop()
+	call.Err = log.err
+
+	// Consume all available gas and clean any leftovers
+	if call.Gas != uint64(0) {
+		call.GasUsed = call.Gas
+	}
+	call.GasIn, call.GasCost = uint64(0), uint64(0)
+	call.OutOff, call.OutLen = nil, nil
+
+	// Flatten the failed call into its parent
+	left := this.callStackLength()
+	if left > 0 {
+		if this.callStack[left-1] == nil {
+			this.callStack[left-1] = &InternalCall{}
+		}
+		if len(this.callStack[left-1].calls) == 0 {
```

Comment:
Is there any reason checking the length istead of checking `nil`?
```suggestion
		if this.callStack[left-1].calls == nil {
```

Reply:
Nothing much, just wanted to take a bit more conservative approach.